### DISCUSSION TITLE
SDF to structure: Some issues

### DISCUSSION THREAD

**Discussion by jrkarki** [rdkit/rdkit#5982] [Q&A] [answered]
Hi everyone,

I am stuck in a situation where I am trying to generate structures for some sdf files. I suspect the sdf information from source has some issues but I'm not able to pin-point it. Btw, instead of reading a sdf file, I am getting the info directly from a json file. 

Below is what I am doing.

my_sdf =  ''\r\n \r\n   -ISIS-  11162210242D\r\n \r\n  28 30  0  0  0  0  0  0  0  0999 V2000\r\n     2.9542   -1.2917    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0\r\n     3.5125   -1.9042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     3.3667   -0.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     4.2625   -1.5667    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0\r\n     4.1792   -0.7417    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0\r\n     2.1417   -1.3792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     0.3542   -0.1292    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0\r\n     0.8375   -0.7917    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     1.6542   -0.7042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     3.3375   -2.7042    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0\r\n     4.9750   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n    -1.1458   -1.7167    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\r\n    -0.3333   -1.6292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     3.0375    0.1833    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\r\n     0.5000   -1.5417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     1.8042   -2.1292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     0.9875   -2.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n    -0.0875    0.4833    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\r\n     0.5750    0.7083    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\r\n    -0.5083   -0.0792    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\r\n     5.6875   -1.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     4.9750   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     4.5875   -0.0292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     5.0000   -0.7417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     6.4042   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     5.6875   -3.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     6.4042   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\r\n     7.1167   -3.2167    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\r\n   2  1  1  0  0  0  0\r\n   3  1  1  0  0  0  0\r\n   4  2  1  0  0  0  0\r\n   5  3  1  0  0  0  0\r\n   6  1  1  0  0  0  0\r\n   7  8  1  0  0  0  0\r\n   8  9  1  0  0  0  0\r\n   9  6  2  0  0  0  0\r\n  10  2  2  0  0  0  0\r\n  11  4  1  0  0  0  0\r\n  12 13  3  0  0  0  0\r\n  13 15  1  0  0  0  0\r\n  14  3  2  0  0  0  0\r\n  15 17  1  0  0  0  0\r\n  16  6  1  0  0  0  0\r\n  17 16  2  0  0  0  0\r\n  18  7  1  0  0  0  0\r\n  19  7  1  0  0  0  0\r\n  20  7  1  0  0  0  0\r\n  21 11  1  0  0  0  0\r\n  22 11  2  0  0  0  0\r\n  23  5  1  0  0  0  0\r\n  24  5  1  0  0  0  0\r\n  25 26  2  0  0  0  0\r\n  26 22  1  0  0  0  0\r\n  27 21  2  0  0  0  0\r\n  28 25  1  0  0  0  0\r\n   4  5  1  0  0  0  0\r\n   8 15  2  0  0  0  0\r\n  27 25  1  0  0  0  0\r\n M  END''

sdf2str = Chem.MolToMolBlock(my_sdf)

I get an error: Counts line too short: ' ' on line4

What I tried next is use the SMILES for the same molecule, convert it to sdf and then try to get the structure. This works fine.

Here's what I did. 

mol = ''FC(F)(F)c1cc(ccc1C#N)N3C(=O)C(C)(C)N(c2ccc(O)cc2)C3=S'
mol_smiles = Chem.MolFromSmiles(mol)
mol2sdf = Chem.MolToMolBlock(mol_smiles)
Chem.MolFromMolBlock(mol2sdf )

where,

mol2sdf = '\n     RDKit          2D\n\n 28 30  0  0  0  0  0  0  0  0999 V2000\n    4.5000    0.0000    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\n    3.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    3.0000   -1.5000    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\n    3.0000    1.5000    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\n    1.5000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    0.7500   -1.2990    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -0.7500   -1.2990    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -1.5000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -0.7500    1.2990    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    0.7500    1.2990    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    1.5000    2.5981    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    2.2500    3.8971    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n   -1.5000   -2.5981    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n   -0.8899   -3.9684    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    0.5773   -4.2803    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\n   -2.0046   -4.9721    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -0.8899   -5.9758    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -2.8863   -6.1856    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -3.3037   -4.2221    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n   -4.6740   -4.8322    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -5.8875   -3.9505    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -7.2578   -4.5606    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -7.4146   -6.0524    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -8.7849   -6.6625    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\n   -6.2011   -6.9341    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -4.8308   -6.3240    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -2.9918   -2.7549    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   -3.9955   -1.6402    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0\n  1  2  1  0\n  2  3  1  0\n  2  4  1  0\n  2  5  1  0\n  5  6  2  0\n  6  7  1  0\n  7  8  2  0\n  8  9  1  0\n  9 10  2  0\n 10 11  1  0\n 11 12  3  0\n  7 13  1  0\n 13 14  1  0\n 14 15  2  0\n 14 16  1  0\n 16 17  1  0\n 16 18  1  0\n 16 19  1  0\n 19 20  1  0\n 20 21  2  0\n 21 22  1  0\n 22 23  2  0\n 23 24  1  0\n 23 25  1  0\n 25 26  2  0\n 19 27  1  0\n 27 28  2  0\n 10  5  1  0\n 26 20  1  0\n 27 13  1  0\nM  END\n'

my_sdf and mol2sdf look similar except that my_sdf has additional '\r' here and there. Could this be the issue? Can anyone try to explain me where the problem is with my_sdf. 

Thanks in advance,
Reagon

**Comment by ptosco** [comment, marked as answer]
@jrkarki `\r\n` only depend on the fact that the file comes from a Windows platform, which uses CR+LF to encode newlines.
The problem with your SDF string is that multiple spaces were collapsed to a single space, thus generating an unreadable molblock.
Once printed, your SDF molblock looks like this:
```

 
 -ISIS- 11162210242D
 
 28 30 0 0 0 0 0 0 0 0999 V2000
 2.9542 -1.2917 0.0000 N 0 0 3 0 0 0 0 0 0 0 0 0
 3.5125 -1.9042 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 3.3667 -0.5667 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 4.2625 -1.5667 0.0000 N 0 0 3 0 0 0 0 0 0 0 0 0
 4.1792 -0.7417 0.0000 C 0 0 3 0 0 0 0 0 0 0 0 0
 2.1417 -1.3792 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 0.3542 -0.1292 0.0000 C 0 0 3 0 0 0 0 0 0 0 0 0
 0.8375 -0.7917 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 1.6542 -0.7042 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 3.3375 -2.7042 0.0000 S 0 0 0 0 0 0 0 0 0 0 0 0
 4.9750 -1.9792 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 -1.1458 -1.7167 0.0000 N 0 0 0 0 0 0 0 0 0 0 0 0
 -0.3333 -1.6292 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 3.0375 0.1833 0.0000 O 0 0 0 0 0 0 0 0 0 0 0 0
 0.5000 -1.5417 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 1.8042 -2.1292 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 0.9875 -2.2167 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 -0.0875 0.4833 0.0000 F 0 0 0 0 0 0 0 0 0 0 0 0
 0.5750 0.7083 0.0000 F 0 0 0 0 0 0 0 0 0 0 0 0
 -0.5083 -0.0792 0.0000 F 0 0 0 0 0 0 0 0 0 0 0 0
 5.6875 -1.5667 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 4.9750 -2.8042 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 4.5875 -0.0292 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 5.0000 -0.7417 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 6.4042 -2.8042 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 5.6875 -3.2167 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 6.4042 -1.9792 0.0000 C 0 0 0 0 0 0 0 0 0 0 0 0
 7.1167 -3.2167 0.0000 O 0 0 0 0 0 0 0 0 0 0 0 0
 2 1 1 0 0 0 0
 3 1 1 0 0 0 0
 4 2 1 0 0 0 0
 5 3 1 0 0 0 0
 6 1 1 0 0 0 0
 7 8 1 0 0 0 0
 8 9 1 0 0 0 0
 9 6 2 0 0 0 0
 10 2 2 0 0 0 0
 11 4 1 0 0 0 0
 12 13 3 0 0 0 0
 13 15 1 0 0 0 0
 14 3 2 0 0 0 0
 15 17 1 0 0 0 0
 16 6 1 0 0 0 0
 17 16 2 0 0 0 0
 18 7 1 0 0 0 0
 19 7 1 0 0 0 0
 20 7 1 0 0 0 0
 21 11 1 0 0 0 0
 22 11 2 0 0 0 0
 23 5 1 0 0 0 0
 24 5 1 0 0 0 0
 25 26 2 0 0 0 0
 26 22 1 0 0 0 0
 27 21 2 0 0 0 0
 28 25 1 0 0 0 0
 4 5 1 0 0 0 0
 8 15 2 0 0 0 0
 27 25 1 0 0 0 0
 M END
```

while it should look like this:
```

  -ISIS-  11162210242D
 
 28 30  0  0  0  0  0  0  0  0999 V2000
    2.9542   -1.2917    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0
    3.5125   -1.9042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.3667   -0.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.2625   -1.5667    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0
    4.1792   -0.7417    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0
    2.1417   -1.3792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.3542   -0.1292    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0
    0.8375   -0.7917    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.6542   -0.7042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.3375   -2.7042    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    4.9750   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.1458   -1.7167    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3333   -1.6292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.0375    0.1833    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.5000   -1.5417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.8042   -2.1292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.9875   -2.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0875    0.4833    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
    0.5750    0.7083    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
   -0.5083   -0.0792    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
    5.6875   -1.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.9750   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.5875   -0.0292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.0000   -0.7417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.4042   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.6875   -3.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.4042   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.1167   -3.2167    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  1  1  0  0  0  0
  4  2  1  0  0  0  0
  5  3  1  0  0  0  0
  6  1  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9  6  2  0  0  0  0
 10  2  2  0  0  0  0
 11  4  1  0  0  0  0
 12 13  3  0  0  0  0
 13 15  1  0  0  0  0
 14  3  2  0  0  0  0
 15 17  1  0  0  0  0
 16  6  1  0  0  0  0
 17 16  2  0  0  0  0
 18  7  1  0  0  0  0
 19  7  1  0  0  0  0
 20  7  1  0  0  0  0
 21 11  1  0  0  0  0
 22 11  2  0  0  0  0
 23  5  1  0  0  0  0
 24  5  1  0  0  0  0
 25 26  2  0  0  0  0
 26 22  1  0  0  0  0
 27 21  2  0  0  0  0
 28 25  1  0  0  0  0
  4  5  1  0  0  0  0
  8 15  2  0  0  0  0
 27 25  1  0  0  0  0
M  END
```
When properly formatted, your SDF block can be parsed by RDKit:
```
sdf ="""
  -ISIS-  11162210242D
 
 28 30  0  0  0  0  0  0  0  0999 V2000
    2.9542   -1.2917    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0
    3.5125   -1.9042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.3667   -0.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.2625   -1.5667    0.0000 N   0  0  3  0  0  0  0  0  0  0  0  0
    4.1792   -0.7417    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0
    2.1417   -1.3792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.3542   -0.1292    0.0000 C   0  0  3  0  0  0  0  0  0  0  0  0
    0.8375   -0.7917    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.6542   -0.7042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.3375   -2.7042    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    4.9750   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.1458   -1.7167    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3333   -1.6292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.0375    0.1833    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.5000   -1.5417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.8042   -2.1292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.9875   -2.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0875    0.4833    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
    0.5750    0.7083    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
   -0.5083   -0.0792    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
    5.6875   -1.5667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.9750   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.5875   -0.0292    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.0000   -0.7417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.4042   -2.8042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.6875   -3.2167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.4042   -1.9792    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.1167   -3.2167    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  1  1  0  0  0  0
  4  2  1  0  0  0  0
  5  3  1  0  0  0  0
  6  1  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9  6  2  0  0  0  0
 10  2  2  0  0  0  0
 11  4  1  0  0  0  0
 12 13  3  0  0  0  0
 13 15  1  0  0  0  0
 14  3  2  0  0  0  0
 15 17  1  0  0  0  0
 16  6  1  0  0  0  0
 17 16  2  0  0  0  0
 18  7  1  0  0  0  0
 19  7  1  0  0  0  0
 20  7  1  0  0  0  0
 21 11  1  0  0  0  0
 22 11  2  0  0  0  0
 23  5  1  0  0  0  0
 24  5  1  0  0  0  0
 25 26  2  0  0  0  0
 26 22  1  0  0  0  0
 27 21  2  0  0  0  0
 28 25  1  0  0  0  0
  4  5  1  0  0  0  0
  8 15  2  0  0  0  0
 27 25  1  0  0  0  0
M  END
"""
mol = Chem.MolFromMolBlock(sdf)
mol
```
![image](https://user-images.githubusercontent.com/5244385/212705015-05922d2d-aa5f-45d2-854a-fce88bfc93c1.png)

**Reply by jrkarki** [reply]
Hi Paolo,

Thanks for your input. I really appreciate it. You mentioned that it is a formatting issue. You seem to have formatted it properly. Can you please explain to me how you formatted it?

Thanks,
Reagon

**Reply by ptosco** [reply]
I reformatted it manually using a Python script, but I would not attempt going down that route - I only did it to demonstrate the problem. You should rather try to get hold of the SDF file in its original format including original spacing rather than trying to reconstruct the original format after spaces were collapsed.
The MDL V2000 format is a fixed format, and there are [many different fields and tags in the MDL specification](http://help.accelrysonline.com/ulm/onelab/1.0/content/ulm_pdfs/direct/reference/ctfileformats2016.pdf), each with its specific fixed spacing. Attempting to fix all of those _a posteriori_ is a significant amount of work. Additionally, you will never be able to recover the original spacing in the SDF properties, since that is entirely dependent on the original data and does not obey to any rules.

**Reply by jrkarki** [reply]
Thanks a lot for helping me out.

**Reply by jrkarki** [reply]
Hi Paolo, 

My fix was to do the following:

my_sdf= my_sdf.replace(' \r\n','\r\n')
my_sdf= my_sdf.replace('\r\n ','\r\n')

However, I will be getting properly formatted data from my data provider. 

Thanks for you help.

Best,
Reagon
